sched/adaptdl_sched/allocator.py

```diff
@@ -38,27 +38,93 @@ class AdaptDLAllocator(object):
     def __init__(self, expander):
         self._core_api = kubernetes.client.CoreV1Api()
         self._objs_api = kubernetes.client.CustomObjectsApi()
+        self._custom_resource = ("adaptdl.petuum.com", "v1",
+                                 "", "adaptdljobs")
         self._cluster_expander = expander
         self._policy = PolluxPolicy()
+        # lock for the two corountines in run()
+        self.lock = asyncio.Lock()
 
     async def run(self):
+        # two functionality: (1) watch for new job and start if possible.
+        # (2) periodically optimize existing jobs
+        await asyncio.gather(
+            self._allocate_one_loop(),
+            self._optimize_all_loop()
+        )
+
+    async def _allocate_one_loop(self):
+        async with kubernetes.watch.Watch() as watch:
+            while True:
+                async for event in watch.stream(
+                        self._objs_api.list_namespaced_custom_object,
+                        *self._custom_resource, timeout_seconds=60):
+                    if event["type"] == "ADDED":  # there is a n arriving job
+                        async with self.lock:
+                            await self._allocate_one(event)
+
+    async def _allocate_one(self, event):
+        # re-read the job , compared with the previous readjob
+        job = event["object"]
+        namespace = job["metadata"]["namespace"]
+        name = job["metadata"]["name"]
+
+        job = await self._objs_api.get_namespaced_custom_object(
```

Comment:
Need to handle 404 errors like is done in the controller, since the job may be deleted at any time. Maybe create an additional method in sched.utils to handle this for both places.

Reply:
added individually. Since the one in controller returns different things.